### PR TITLE
fix: text overflow & portfolio hr

### LINKED ISSUE #37141: [Beta] Portfolio description is not wrapping at the boundary
**Describe the bug**
A long description of a portfolio item doesn't wrap at the boundary.
Screenshot (25)

**To Reproduce**
1. Go to the settings page
2. Put a long description in the portfolio item
3. Check out the public portfolio (hard refresh it ctrl+f5; this update bug is under triage)

**Expected behavior**
The description should wrap at the boundary of div.

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro 32 Bit Pro
 - Browser: Chrome
 - Version: 77.0.3865.90

### PR DESCRIPTION
**Pull request by vkWeb**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #37141 

**This PR addresses the following concerns:**
- [x] Now the word wraps at the boundary in portfolio title, description, username, name, about and flash messages.
- [x] Portfolio hr is now consistent with global hr.
- [x] Textarea horizontal resize has been disabled to prevent textarea horizontal overflow.

### CODE REVIEW

**Comment by vkWeb** [comment]
Closes #37209 as well.

**Comment by ojeytonwilliams** [comment]
@vkWeb thanks for the PR.  

I'm a little concerned that these selectors might be a bit too general and have unintended consequences.  For instance, the wallets have their own class `wallet` that could be used to break the words up, rather than breaking all code elements.

To be clear, I'm not saying this has broken anything, I'm just not sure using broad selectors to deal with a localised problem is the way to go.

Also, any idea why `overflow-wrap: break-word;` stopped working?  It used to: https://github.com/freeCodeCamp/freeCodeCamp/pull/36688

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams like you read my mind :)

**Comment by vkWeb** [comment]
@ojeytonwilliams @ahmadabdolsaheb Yes you both are right, I was also not inclined to take this global selector path. But I observed overflow in almost every place in the app. Here are some screenshots of the overflows:

Screenshot (30)

Screenshot (31)

Screenshot (27)

Screenshot (28)

Screenshot (29)

So instead of targeting each page/component, I decided to declare global selectors. I think it won't have any unintended consequences as we need to break-word everywhere in the app, right?

> Also, any idea why overflow-wrap: break-word; stopped working? It used to: #36688

@ahmadabdolsaheb Any idea why is this happening? I searched the internet for overflow-wrap and word-break but couldn't find a satisfactory difference between them.

**Comment by ojeytonwilliams** [comment]
Thanks for the screenshots - it's more widespread than I realised. I see the justification for those selectors now.

What I would recommend is that either the wallets are broken using `.wallet` or they are wrapped in something other than `code`. Either way, I don't think code should allow any breaks, since that can change it's meaning.

**Comment by vkWeb** [comment]
> Thanks for the screenshots - it's more widespread than I realised. I see the justification for those selectors now.

Yes :)

> What I would recommend is that either the wallets are broken using .wallet or they are wrapped in something other than code. 

https://github.com/freeCodeCamp/freeCodeCamp/blob/372cba1c4cd6ae63fc17d385d1d6ffff1556ddbb/client/src/components/Donation/components/DonateOther.js#L165-L170

It is wrapped in `code` with className `wallet`. 

>Either way, I don't think code should allow any breaks, since that can change it's meaning.

@ojeytonwilliams we will need to do it to prevent overflow and why will it change its meaning? We are breaking by word that means, words that can't fit in the width will break to newline and if they still can't fit in then only it will get a break at an arbitrary point. I hope I haven't confused you 😅.

**Comment by ojeytonwilliams** [comment]
> @ojeytonwilliams we will need to do it to prevent overflow and why will it change its meaning? We are breaking by word that means, words that can't fit in the width will break to newline and if they still can't fit in then only it will get a break at an arbitrary point. I hope I haven't confused you .

Not at all.  My concern was that linebreaks (in many programming languages) can change what the code does or even if it is valid syntax.  However, fenced code already prevents wrapping (it scrolls instead), so there's only a real problem if a challenge has a lot of text in a code element.  @RandellDawson do you think you could come up with a regex to flag up any files with overly long `<code>` blocks?  No particular rush.  If there are any issues we can fix them as and when we find them.

As an example of why this can matter, consider that

```js
function longFunction
Name(x) {console.log(
x)}
```

is not valid JavaScript while
```js
function longFunctionName(x) {console.log(x)}
```
is fine.  

Anyway, that's the only situation I think line or word breaks would be bad, so I'm actually going to reverse my original position.  How about we just use

```css
body {
  word-break: break-word
}
```

It seems to be that (aside from code) it's almost always going to be better to break rather than overflow, given how annoying this is on mobile.  @ahmadabdolsaheb what do you think?

**Comment by ahmadabdolsaheb** [comment]
There are other instances where we don't want the code to wrap such as one line comments.

For the wallets, is there any specific reason why we need to use the <code> there?

Also since we are really close to shipping master ( a couple of days ) I would rather fix those wrap instances on a case by case basis. only because we would not have enough time to audit the whole app again.

On the next design iteration we would push for removing bootstrap and incorporating our own component library, and grid system. then we could normalize the styles accordingly. 

So let's resolve those issues with specific styles if we could, if not I am ok with setting more general styles if we have to.

**Comment by vkWeb** [comment]
>There are other instances where we don't want the code to wrap such as one-line comments.

Yes, I agree that putting word-break in the body will not be a good idea, it can interfere with local styles.

>Also since we are really close to shipping master ( a couple of days ) I would rather fix those wrap instances on a case by case basis. only because we would not have enough time to audit the whole app again.

The tags in which I have applied the style will need word-break in all cases so it won't harm us. We can do it case-by-case but I won't recommend doing it because putting that same trivial style in a lot of places won't be a good idea. 

>On the next design iteration we would push for removing bootstrap and incorporating our own component library, and grid system. then we could normalize the styles accordingly.

@ahmadabdolsaheb I would love to give you a helping hand on the next design iteration :).

>So let's resolve those issues with specific styles if we could, if not I am ok with setting more general styles if we have to.

I think we should merge this as is.

**Comment by ojeytonwilliams** [comment]
Good point about having to re-audit, I wasn't thinking straight.

I don't think wallets need to be inside `<code>`, I think it's mostly to highlight them.

**Comment by ojeytonwilliams** [comment]
> Yes, I agree that putting word-break in the body will not be a good idea, it can interfere with local styles.

The structure of the site is basically

```html
<body>
  <div>
    stuff
  </div>
</body>
```
so 
```css
body {
  word-break: break-word;
}
```
and
```css
div {
  word-break: break-word;
}
```

are basically equivalent.  You can see this by just using the second rule and you should see that everything that was overflowing now breaks.

All I'm getting at is, if we were going to apply it globally (and we probably shouldn't right now for the reasons that @ahmadabdolsaheb went through), then we might as well make it obvious that it's meant to be global.  

The ideal solution is to use the classes like `big-heading` `wallet` and `username` to apply `word-break: break-word` to make limited changes that we can QA.  Later on we can simply revert this and apply `word-break: break-word` globally when we have time to ensure that that doesn't create new problems.

**Comment by vkWeb** [comment]
@ojeytonwilliams Sounds like a plan. I will fix overflows case by case for now. On the next iteration, we'll think of handling this issue more globally.

Pushing soon.

**Comment by ojeytonwilliams** [comment]
Thanks @vkWeb. Sorry for pulling you in the wrong direction on this one!

**Comment by vkWeb** [comment]
@ojeytonwilliams No problem man ;). The refactor isn't a big one, it's all easy.

**Comment by vkWeb** [comment]
@ahmadabdolsaheb @ojeytonwilliams Before we merge this, please keep in mind that `word-break: break-word` is a deprecated API. It still works on almost all browsers but we should remove it in future design iterations. 

Hence I have used `overflow-wrap: break-word` in all the places it was possible. `overflow-wrap: break-word` only works when the width is specified but `word-break: break-word` works in all the cases. So in the case of `wallet` and `media` (portfolio), I have used `word-break` because `overflow-wrap` wasn't working. 

This is ready to merge.

File: client/src/components/profile/components/camper.css [hunk 1]
```diff
@@ -6,3 +6,9 @@
   display: flex;
   justify-content: center;
 }
+
+.username,
+.name,
```
Comment (ojeytonwilliams) [inline client/src/components/profile/components/camper.css]:
```suggestion
.name,
.location,
```
`location` can also be arbitrarily long.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Looks great, @vkWeb!  I checked all the places you have allowed breaks and they're much better now.  

I found one case, `location`, that could also cause overflow, so if you could fix that I think we're good to go.

File: client/src/components/profile/components/camper.css (hunk 1 above)
Comment (vkWeb) [reply client/src/components/profile/components/camper.css]: @ojeytonwilliams Good catch. Pushing in a sec.

**Comment by vkWeb** [comment]
@ojeytonwilliams something went wrong. Reopening, resolving and pushing soon.

**Comment by vkWeb** [comment]
@ojeytonwilliams I hope things look good now.

**Comment by vkWeb** [comment]
@ojeytonwilliams ready to merge.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

@ahmadabdolsaheb this deals with all the overflow I could find and the fixes are local.

**Comment by ahmadabdolsaheb** [comment]
>@ahmadabdolsaheb I would love to give you a helping hand on the next design iteration :).
let's do it :)

looks good, there is one tiny little thing. 

I appreciate it if you could finalize this today, if not we could address it in another issue and fix it later.

Screen Shot 2019-10-15 at 11 19 49 AM

**Comment by vkWeb** [comment]
@ahmadabdolsaheb I'll fix it today.

**Comment by vkWeb** [comment]
@ahmadabdolsaheb Done.

**Comment by ahmadabdolsaheb** [comment]
@vkWeb I really appreciate all the efforts you put into this pull request.

Since we have shipped master and are not in a rush, I was wondering if we could implement a more general solution similar to your initial approach or what @ojeytonwilliams suggested with the styling the body.

Apologies for the inconvenience.

**Comment by vkWeb** [comment]
@ahmadabdolsaheb No problem at all ;). I am ready to make the platform as resilient as possible. Can you suggest me a good approach we should take?

**Comment by ojeytonwilliams** [comment]
I'm not averse to merging this, with the understanding that we're going to overhaul it at some point.  After all, we've all checked it and are happy that it handles the overflow issues.  It just depends on how soon we're likely to be pushing changes to master.

As to what's next: I think, ideally, we'd want to use `overflow-wrap: word-break` globally and audit the site to make sure that doesn't break anything.  However, @vkWeb noticed that you need `word-break: break-word` in several cases.  I think that's due to a weird interaction between `display: flex` and `margin: auto`.

I put together an extremely ugly pen that demonstrates this: https://codepen.io/ojeytonwilliams/pen/wvvGmRM
You can see that the text overflows the red box despite having `overflow-wrap: word-break` and that it stays within the yellow one (which uses `margin-left:auto`).  @ahmadabdolsaheb you'll not be surprised to learn that this another issue with Bootstrap.

If I recall correctly we're planning to move away from Bootstrap, so perhaps we could revisit this problem when that migration happens?

I'm mostly thinking out-loud here, so apologies if I've entirely missed the mark.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams sounds good.
Let's merge this in, and revisit it during the overhaul.

**Comment by vkWeb** [comment]
Hey @ojeytonwilliams thanks for the explanation. Now I understood why `overflow-wrap: break-word` wasn't working. 

@ahmadabdolsaheb Yes we'll revisit this later.

**Comment by ojeytonwilliams** [comment]
@vkWeb you're welcome.  It was bugging me, so I felt like I had to investigate...

**Pull request merged**